lcls-twincat-common-components/Library/Devices/XPIM/FB_XPIM.TcPOU

```diff
@@ -0,0 +1,231 @@
+    // The current position state as an enum.
+    {attribute 'pytmc' := '
+        pv: MMS:STATE:GET
+        io: i
+    '}
+    eEnumGet: E_XPIM_States;
+    // The PMPS database lookup associated with the current position state.
+    stDbStateParams: ST_DbStateParams;
+END_VAR
+VAR
+    bInit: BOOL;
+
+    fbYStage: FB_MotionStage;
+    fbZoom: FB_MotionStage;
+    fbFocus: FB_MotionStage;
+
+    fbStateDefaults: FB_PositionState_Defaults;
+
+    {attribute 'pytmc' := '
+        pv: MMS
+        astPositionState.array: 1..4
+    '}
+    fbStates: FB_PositionStatePMPS1D;
+    astPositionState: ARRAY[1..GeneralConstants.MAX_STATES] OF ST_PositionState;
+    fbArrCheckWrite: FB_CheckPositionStateWrite;
+
+    {attribute 'pytmc' := 'pv: MFW'}
+    fbFilterWheel: FB_XPIM_FilterWheel;
+
+    {attribute 'pytmc' := 'pv: CAM'}
+    fbOpal: FB_XPIM_Opal;
+
+    {attribute 'pytmc' := 'pv: CIL'}
+    fbLED: FB_XPIM_LED;
+
+    {attribute 'pytmc' := 'pv: SFW'}
+    fbFlowSwitch: FB_XTES_Flowswitch;
+END_VAR
+VAR CONSTANT
+    // State defaults if not provided
+    fDelta: LREAL := 2;
+    fAccel: LREAL := 200;
+    fOutDecel: LREAL := 25;
+END_VAR]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[
+IF NOT bInit THEN
+    bInit := TRUE;
+
+    stYStage.bHardwareEnable := TRUE;
+    stYStage.bPowerSelf := FALSE;
+    // No limit switch at the bottom
+    stYStage.bLimitBackwardEnable := TRUE;
+
+    // Extra lock on lens + lens limits are NO instead of NC
+    stZoomStage.bHardwareEnable := NOT bZoomLock;
+    stZoomStage.bPowerSelf := TRUE;
+    stZoomStage.bLimitForwardEnable := NOT bZoomEndFwd;
+    stZoomStage.bLimitBackwardEnable := NOT bZoomEndBwd;
+    stZoomStage.nHomingMode := ENUM_EpicsHomeCmd.LOW_LIMIT;
+    stZoomStage.fHomePosition := 0;
+
+    stFocusStage.bHardwareEnable := NOT bFocusLock;
+    stFocusStage.bPowerSelf := TRUE;
+    stFocusStage.bLimitForwardEnable := NOT bFocusEndFwd;
+    stFocusStage.bLimitBackwardEnable := NOT bFocusEndBwd;
+    stFocusStage.nHomingMode := ENUM_EpicsHomeCmd.LOW_LIMIT;
+    stFocusStage.fHomePosition := 0;
+
+    // Partial backcompat, this used to set fVelocity too but this should be set per install
+    fbStateDefaults(stPositionState:=stOut, sNameDefault:='OUT', fDeltaDefault:=fDelta, fAccelDefault:=fAccel, fDecelDefault:=fOutDecel);
+    fbStateDefaults(stPositionState:=stYag, sNameDefault:='YAG', fDeltaDefault:=fDelta, fAccelDefault:=fAccel, fDecelDefault:=fAccel);
+    fbStateDefaults(stPositionState:=stDiamond, sNameDefault:='DIAMOND', fDeltaDefault:=fDelta, fAccelDefault:=fAccel, fDecelDefault:=fAccel);
+    fbStateDefaults(stPositionState:=stReticle, sNameDefault:='RETICLE', fDeltaDefault:=fDelta, fAccelDefault:=fAccel, fDecelDefault:=fAccel);
+END_IF
+
+fbYStage(stMotionStage:=stYStage);
+fbZoom(stMotionStage:=stZoomStage);
+fbFocus(stMotionStage:=stFocusStage);
+
+// Set special error message for lens lock
+IF stZoomStage.bExecute AND bZoomLock THEN
+    IF NOT stZoomStage.bError THEN
```

Comment:
Is this `IF` really necessary? Seems like you could set it always just as the custom message

Reply:
I think the original intent was "if there isn't already some other, more specific error" but it escapes me how the other error could ever be more specific.

Reply:
Looking more closely at the expanded file here, I no longer stand by past me's decision. It must simply be a brain moment.